### PR TITLE
Reachable smart-tag and metadata redesign. Predicate update and assembly

### PR DESCRIPTION
**Pull request by adlrocha**
Implements: #30 
Fixes bugs and makes improvements.

### Reachable re-design
- [x] Tweak design of "reachable" smart tag
- [x] Decouple verification of reachability from the assembly process
- [x] Redesign Reachable 
  - [x] Add two predicates for Reachable: `connectivity` and `dialability`.
  - [x] Update after check to: `dialled` (if dials), `connected` (if connects), `notDialable` (not reachable), and `notConnected`.
  - [x] Disassembly to predicates with the right tags.
- [x] Tweak the assembler for reachable to just fill in the above structure (with verifiedXXX = false)
- [x] Modify the VM's Update to follow the above

### IR Predicate updates
- [x] Implement predicate.UpdateWith(update)
- [x] Predicate Assembler

### Metadata re-design
- [x] Re-design metadata so it can be used by smart-tags.

### UpdateWith re-design
Fix: https://github.com/libp2p/go-smart-record/issues/32
`UpdateWiith` in basic types overwrites the previous value as long as it is the same type.
To overwrite for other type, `Set` function will be implemented.

### CODE REVIEW

**Comment by adlrocha** [comment]
Hey, @petar. After our discussion from yesterday, this is ready for review. Thanks!

File: ir/base/reachable.go [hunk 1]
```diff
@@ -70,166 +84,128 @@ func (r *Reachable) UpdateWith(ctx ir.UpdateContext, with ir.Node) error {
...
-		return nil, fmt.Errorf("expecting tag 'connected' or 'dialable'")
-	}
-	// If the node is of type connected List flag
-	if tag == "connected" {
-		isConn = true
+// TriggerReachable triggers the execution of Reachable verifications
+// over a dict and adds the appropiate flag to Nodes that don't pass the verification.
+func TriggerReachable(d *ir.Dict, h host.Host) {
+	// For each pair.
+	for k := len(d.Pairs) - 1; k >= 0; k-- {
```
Comment (petar) [inline ir/base/reachable.go:153]: Why is the iteration going backwards? Couldn't you just use:
for _, p := d.Pairs {
  ...
}

**Review by petar — APPROVED** [review]
LGTM. One small comment.

File: ir/base/reachable.go (hunk 1 above)
Comment (adlrocha) [reply ir/base/reachable.go]: Good point! In the previous implementation, non-reachable nodes were being removed in each iteration (and thus the backward loop). Now we don't need it but I forgot to change it. Now is fixed. Thanks!

**Pull request merged**